src/ring_theory/unique_factorization_domain.lean

```diff
@@ -248,40 +224,109 @@ by haveI := classical.dec_eq α; exact
         (hf p (by simp)).ne_zero))
     end)
 
-end unique_factorization_domain
+end unique_factorization_monoid
 
-structure unique_irreducible_factorization (α : Type*) [integral_domain α] :=
-(factors : α → multiset α)
-(factors_prod : ∀{a : α}, a ≠ 0 → (factors a).prod ~ᵤ a)
-(irreducible_factors : ∀{a : α}, a ≠ 0 →  ∀x∈factors a, irreducible x)
-(unique : ∀{f g : multiset α},
-  (∀x∈f, irreducible x) → (∀x∈g, irreducible x) → f.prod ~ᵤ g.prod → multiset.rel associated f g)
+lemma prime_factors_unique [comm_cancel_monoid_with_zero α] : ∀{f g : multiset α},
+  (∀x∈f, prime x) → (∀x∈g, prime x) → f.prod ~ᵤ g.prod →
+  multiset.rel associated f g :=
+by haveI := classical.dec_eq α; exact
+λ f, multiset.induction_on f
+  (λ g _ hg h,
+    multiset.rel_zero_left.2 $
+      multiset.eq_zero_of_forall_not_mem (λ x hx,
+        have is_unit g.prod, by simpa [associated_one_iff_is_unit] using h.symm,
+        (irreducible_of_prime $ hg x hx).1 (is_unit_iff_dvd_one.2 (dvd.trans (multiset.dvd_prod hx)
+          (is_unit_iff_dvd_one.1 this)))))
+  (λ p f ih g hf hg hfg,
+    let ⟨b, hbg, hb⟩ := exists_associated_mem_of_dvd_prod
+      (hf p (by simp)) (λ q hq, hg _ hq) $
+        (dvd_iff_dvd_of_rel_right hfg).1
+          (show p ∣ (p :: f).prod, by simp) in
+    begin
+      rw ← multiset.cons_erase hbg,
+      exact multiset.rel.cons hb (ih (λ q hq, hf _ (by simp [hq]))
+        (λ q (hq : q ∈ g.erase b), hg q (multiset.mem_of_mem_erase hq))
+        (associated_mul_left_cancel
+          (by rwa [← multiset.prod_cons, ← multiset.prod_cons, multiset.cons_erase hbg]) hb
+        (hf p (by simp)).ne_zero))
+    end)
 
-namespace unique_factorization_domain
-open unique_factorization_domain associated
-variables [integral_domain α] [unique_factorization_domain α]
+lemma prime_factors_irreducible [comm_cancel_monoid_with_zero α] {a : α} {f : multiset α}
+  (ha : irreducible a) (pfa : (∀b ∈ f, prime b) ∧ a ~ᵤ f.prod) :
+  ∃ p, a ~ᵤ p ∧ f = p :: 0 :=
+by haveI := classical.dec_eq α; exact
+multiset.induction_on f
+  (λ h, (ha.1 (associated_one_iff_is_unit.1 h)).elim)
+  (λ p s _ hp hs, let ⟨u, hu⟩ := hp in ⟨p,
+    have hs0 : s = 0, from classical.by_contradiction
+      (λ hs0, let ⟨q, hq⟩ := multiset.exists_mem_of_ne_zero hs0 in
+       (hs q (by simp [hq])).2.1 $
+        (ha.2 ((p * ↑(u⁻¹)) * (s.erase q).prod) _
+          (by {rw [mul_right_comm _ _ q, mul_assoc, ← multiset.prod_cons, multiset.cons_erase hq,
+            mul_comm p _, mul_assoc, ← multiset.prod_cons, hu.symm, mul_comm, mul_assoc],
+            simp, })).resolve_left $
+              mt is_unit_of_mul_is_unit_left $ mt is_unit_of_mul_is_unit_left
+                (hs p (multiset.mem_cons_self _ _)).2.1),
+    ⟨(by {clear _let_match; simp * at *}), hs0 ▸ rfl⟩⟩)
+  (pfa.2) (pfa.1)
```

Comment:
While the proof is fresh in your mind you could even add a couple of comments saying what's going on. Christian Szegedy told me today that the machine learning people are desperate for comments in code.

Reply:
I've made this a tactic proof, but I can't really say it's "fresh in my mind", because I didn't write the original and don't particularly understand it. I've added a docstring to the top because it took me a moment even to understand the statement.